crates/swc_ecma_transforms_react/src/refresh/hook.rs

```diff
@@ -0,0 +1,523 @@
+                span: DUMMY_SP,
+                value: should_reset,
+            })));
+        }
+
+        if custom_hook_in_scope.len() > 0 {
+            let elems = custom_hook_in_scope
+                .into_iter()
+                .map(|hook| {
+                    Some(ExprOrSpread {
+                        spread: None,
+                        expr: Box::new(match hook {
+                            HookCall::Ident(ident) => Expr::Ident(ident),
+                            HookCall::Member(obj, prop) => Expr::Member(MemberExpr {
+                                span: DUMMY_SP,
+                                obj: ExprOrSuper::Expr(Box::new(obj)),
+                                prop: Box::new(Expr::Ident(prop)),
+                                computed: false,
+                            }),
+                        }),
+                    })
+                })
+                .collect();
+            args.push(Expr::Fn(FnExpr {
+                ident: None,
+                function: Function {
+                    is_generator: false,
+                    is_async: false,
+                    params: Vec::new(),
+                    decorators: Vec::new(),
+                    span: DUMMY_SP,
+                    body: Some(BlockStmt {
+                        span: DUMMY_SP,
+                        stmts: vec![Stmt::Return(ReturnStmt {
+                            span: DUMMY_SP,
+                            arg: Some(Box::new(Expr::Array(ArrayLit {
+                                span: DUMMY_SP,
+                                elems,
+                            }))),
+                        })],
+                    }),
+                    type_params: None,
+                    return_type: None,
+                },
+            }));
+        }
+
+        Expr::Call(CallExpr {
+            span: DUMMY_SP,
+            callee: ExprOrSuper::Expr(Box::new(Expr::Ident(handle))),
+            args: args
+                .into_iter()
+                .map(|arg| ExprOrSpread {
+                    spread: None,
+                    expr: Box::new(arg),
+                })
+                .collect(),
+            type_args: None,
+        })
+    }
+
+    fn gen_hook_register_stmt(&mut self, ident: Ident, sig: HookSig) {
+        self.ident.push(sig.handle.clone());
+        self.extra_stmt.push(Stmt::Expr(ExprStmt {
+            span: DUMMY_SP,
+            expr: Box::new(self.wrap_with_register(sig.handle, Expr::Ident(ident), sig.hooks)),
+        }))
+    }
+}
+
+impl<'a> VisitMut for HookRegister<'a> {
+    noop_visit_mut_type!();
+
+    fn visit_mut_block_stmt(&mut self, b: &mut BlockStmt) {
+        let mut current_scope = IndexSet::new();
+
+        // TODO: merge with collect_decls
+        for stmt in &b.stmts {
+            stmt.collect_ident(&mut current_scope);
+        }
+        let orig_binding = self.scope_binding.len();
+        self.scope_binding.extend(current_scope.into_iter());
```

Comment:
Is `.into_iter()` required?
`.extend` accepts `impl IntoIterator` so this seems redundant.